code-orange/Communicator.swift

```diff
@@ -8,56 +8,59 @@
 
 import Foundation
 
+protocol DataFetcher {
+  func updateInfectedLocations()
+  func getInfectedLocations() -> [COLocation]
+}
+
 class Communicator {
-  func getServerResults() -> [RecordedLocation]? {
-    let stringData = getFromAPI()
-    guard let data = stringData.data(using: .utf8, allowLossyConversion: false) else {
-      return nil
-    }
-    
-    guard let serverRecorededLocations = decode(data: data) else { return nil }
-    let recordedLocations = serverRecorededLocations.compactMap { RecordedLocation(serverLocation: $0)}
-    return recordedLocations
-  }
-  
-  // TEMP - remove this
-  func getUserData() -> [RecordedLocation]? {
-    let stringData = getFromStorage()
-    guard let data = stringData.data(using: .utf8, allowLossyConversion: false) else {
-      return nil
-    }
-    
-    guard let serverRecorededLocations = decode(data: data) else { return nil }
-    let recordedLocations = serverRecorededLocations.compactMap { RecordedLocation(serverLocation: $0)}
-    return recordedLocations
+  private let session: URLSession
+  private let serverUrl = "http://ec2-52-23-173-222.compute-1.amazonaws.com:8080/v1/events/covid-19/locations?patient_status=carrier&country=il"
+
+  private lazy var jsonDecoder: JSONDecoder = {
+    let jsonDecoder = JSONDecoder()
+    let dateFormatter = DateFormatter()
+    dateFormatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss.SSSZ"
+    jsonDecoder.dateDecodingStrategy = .formatted(dateFormatter)
+    return jsonDecoder
+  }()
+
+  private var infectedLocations = [COLocation]()
+
+  init(session: URLSession = .shared) {
+    self.session = session
+
+    updateInfectedLocations()
   }
-  
-  func decode(data: Data) -> [ServerRecordedLocation]? {
-    let decoder = JSONDecoder()
-    let decodedData: [String: [ServerRecordedLocation]]
-    do {
-      try decodedData = decoder.decode([String: [ServerRecordedLocation]].self, from: data)
-    } catch {
-      print("This Shouldn't happen: \(error)")
-      return nil
+
+  private func notifyConsumers() {
```

Comment:
```suggestion
  private func notifyDownloadCompleted()
```

consumers in a bit vague

Reply:
Done!